### PR TITLE
feat(docdb): add monitoring for Document DB

### LINKED ISSUE #167: Monitoring support for AWS DocumentDB
### Feature scope

DocumentDB

### Describe your suggested feature

I just installed v1.10.0 and it does appear that there isn't support for monitoring AWS DocumentDB via this library? I am kindly requesting that this be considered and added, if it's not already been planned.

### PR DESCRIPTION
**Pull request by voho**
Fixes #167 

---

_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license_

### CODE REVIEW

**Review by echeung-amzn — CHANGES_REQUESTED** [review]
Add to `MonitoringAspect` as well?

File: lib/monitoring/index.ts [hunk 1]
```diff
@@ -6,6 +6,7 @@ export * from "./aws-billing";
 export * from "./aws-cloudfront";
 export * from "./aws-cloudwatch";
 export * from "./aws-codebuild";
+export * from "./aws-documentdb";
```
[the remark this answers is not in the document]
Comment (voho) [reply lib/monitoring/index.ts:9]: OK

File: lib/monitoring/aws-documentdb/DocumentDbMetricFactory.ts [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  getLatencyTypeStatistic,
+  LatencyType,
+  MetricFactory,
+  MetricStatistic,
+} from "../../common";
+
+const DocumentDbNamespace = "AWS/DocDB";
+
+export interface DocumentDbMetricFactoryProps {
+  readonly clusterIdentifier: string;
```
[the remark this answers is not in the document]
Comment (voho) [reply lib/monitoring/aws-documentdb/DocumentDbMetricFactory.ts:14]: If such construct already exists, then I will go for it! I was not aware.

File: lib/monitoring/aws-documentdb/DocumentDbMetricFactory.ts [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    props: DocumentDbMetricFactoryProps
+  ) {
+    this.metricFactory = metricFactory;
+    this.dimensionsMap = {
+      DBClusterIdentifier: props.clusterIdentifier,
+    };
+  }
+
+  metricAverageCpuUsageInPercent() {
+    return this.metric("CPUUtilization", MetricStatistic.AVERAGE, "CPU Usage");
```
[the remark this answers is not in the document]
Comment (voho) [reply lib/monitoring/aws-documentdb/DocumentDbMetricFactory.ts:32]: I have created a burner account with docdb to check the metrics, will verify the datapoints are meaningful - the datapoints should be populated by now.

The `adaptMetric` is done in the private `metric` method, so we should be covered.

File: lib/monitoring/aws-documentdb/DocumentDbMonitoring.ts [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    this.usageAnnotations = [];
+
+    const metricFactory = new DocumentDbMetricFactory(
+      scope.createMetricFactory(),
+      props
+    );
+
+    this.cpuUsageMetric = metricFactory.metricAverageCpuUsageInPercent();
+    this.readLatencyMetric = metricFactory.metricAverageCpuUsageInPercent();
+    this.writeLatencyMetric = metricFactory.metricAverageCpuUsageInPercent();
```
[the remark this answers is not in the document]
Comment (voho) [reply lib/monitoring/aws-documentdb/DocumentDbMonitoring.ts:78]: Oops. Great catch.

**Comment by voho** [comment]
> Add to `MonitoringAspect` as well?

Happy to, since there is this L2 construct. However, I am not sure if there is a way to distinguish different cluster types, e.g. DocDB cluster from RDS cluster. Will check what the exact type of differentiator is.

File: lib/monitoring/aws-docdb/DocumentDbMetricFactory.ts [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+  MetricStatistic,
+} from "../../common";
+
+const DocumentDbNamespace = "AWS/DocDB";
+
+export interface DocumentDbMetricFactoryProps {
+  /**
+   * database cluster identifier (either this or `cluster` need to be specified)
+   */
+  readonly clusterIdentifier?: string;
```
Comment (ayush987goyal) [inline lib/monitoring/aws-docdb/DocumentDbMetricFactory.ts:18]: Do we really need to support both? We typically are not doing that for other monitorings.

File: lib/monitoring/aws-docdb/DocumentDbMetricFactory.ts [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+  ) {
+    return this.metricFactory.createMetric(
+      metricName,
+      statistic,
+      label,
+      this.dimensionsMap,
+      undefined,
+      DocumentDbNamespace
+    );
+  }
```
Comment (ayush987goyal) [inline lib/monitoring/aws-docdb/DocumentDbMetricFactory.ts:120]: Would this need to be wrapped with adapt?

File: lib/monitoring/aws-docdb/DocumentDbMetricFactory.ts [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+      return props.clusterIdentifier;
+    } else if (
+      props.cluster !== undefined &&
+      props.clusterIdentifier === undefined
+    ) {
+      return props.cluster.clusterIdentifier;
+    } else {
+      throw new Error(
+        "Specify either 'clusterIdentifier' or 'cluster', but not both."
+      );
```
Comment (ayush987goyal) [inline lib/monitoring/aws-docdb/DocumentDbMetricFactory.ts:53]: This else block also covers the case where neither of them are specified. The message in that case would be misleading.
We should add two different error messages when both or neither are provided

**Review by ayush987goyal — CHANGES_REQUESTED** [review]

File: lib/monitoring/aws-docdb/DocumentDbMetricFactory.ts (hunk 5 above)
Comment (voho) [reply lib/monitoring/aws-docdb/DocumentDbMetricFactory.ts:18]: I know, but I was thinking about two reasons:

1) The cluster L2 constructs are quite new and users might not be using them yet (they might only have cluster ID available)
2) The other construct (RDS) was using identifier and should be using cluster too - here I wanted to be backwards compatible and consistent

Not sure it is the best decision, what do you think?

File: lib/monitoring/aws-docdb/DocumentDbMetricFactory.ts [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+export interface DocumentDbMetricFactoryProps {
+  /**
+   * database cluster identifier (either this or `cluster` need to be specified)
+   */
+  readonly clusterIdentifier?: string;
+  /**
+   * database cluster (either this or `clusterIdentifier` need to be specified)
+   */
+  readonly cluster?: DatabaseCluster;
```
[the remark this answers is not in the document]
Comment (voho) [reply lib/monitoring/aws-docdb/DocumentDbMetricFactory.ts:22]: Indeed!

File: lib/monitoring/aws-docdb/DocumentDbMetricFactory.ts (hunk 7 above)
Comment (voho) [reply lib/monitoring/aws-docdb/DocumentDbMetricFactory.ts:53]: OK - I thought it was clear enough, but we can raise the bar here and be very specific.

File: lib/monitoring/aws-docdb/DocumentDbMetricFactory.ts (hunk 5 above)
Comment (ayush987goyal) [reply lib/monitoring/aws-docdb/DocumentDbMetricFactory.ts:18]: 1. Even though it is new, if it comes under our minimum supported `aws-cdk-lib` version we should use it. For the users who might have just the ID available, they can import the Cluster via [fromDatabaseClusterAttributes](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_docdb.DatabaseCluster.html#static-fromwbrdatabasewbrclusterwbrattributesscope-id-attrs)  which takes [DatabaseClusterAttributes](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_docdb.DatabaseClusterAttributes.html) which inturn only needs the ID as required argument.
2. For RDS, we can keep it the way you have right now (supporting both) but we should mark the ID as depreacated and instruct them to use the IDatabaseCluster instead (same things apply here)

Comment (voho) [reply lib/monitoring/aws-docdb/DocumentDbMetricFactory.ts:18]: OK, no problem, I can do it like this too.

File: lib/monitoring/aws-docdb/DocumentDbMetricFactory.ts (hunk 7 above)
Comment (voho) [reply lib/monitoring/aws-docdb/DocumentDbMetricFactory.ts:53]: cool, will do

**Review by ayush987goyal — APPROVED** [review]
Thanks a lot for making these changes!

**Comment by voho** [comment]
Hey folks, thanks for review. Did I miss any change requested? Please let me know :).

**Review by echeung-amzn — APPROVED** [review]

**Pull request merged**